src/main/java/jenkins/plugins/git/GitUsernamePasswordBinding.java

```diff
@@ -0,0 +1,182 @@
+    private String gitToolExe = null;
+    private boolean unixNodeType;
+
+    @DataBoundConstructor
+    public GitUsernamePasswordBinding(String gitToolName, String credentialsId) {
+        super(credentialsId);
+        this.gitToolName = gitToolName;
+        //Variables could be added if needed
+    }
+
+    public String getGitToolName(){
+        return this.gitToolName;
+    }
+
+    private void setUnixNodeType(boolean value) {
+        this.unixNodeType = value;
+    }
+
+    @Override
+    protected Class<StandardUsernamePasswordCredentials> type() {
+        return StandardUsernamePasswordCredentials.class;
+    }
+
+    @Override
+    public MultiEnvironment bind(@NonNull Run<?, ?> run, FilePath filePath,
+                                 Launcher launcher, @NonNull TaskListener taskListener)
+            throws IOException, InterruptedException {
+        StandardUsernamePasswordCredentials credentials = getCredentials(run);
+        setKeyBindings(credentials);
+        gitToolExe = getGitTool(run, this.gitToolName, taskListener);
+        setUnixNodeType(isCurrentNodeOSUnix(launcher));
+        if (gitToolExe != null && filePath != null) {
+            final UnbindableDir unbindTempDir = UnbindableDir.create(filePath);
+            setRunEnvironmentVariables(filePath, taskListener);
+            GenerateGitScript gitScript = new GenerateGitScript(
+                                                  credentials.getUsername(),
+                                                  credentials.getPassword().getPlainText(),
+                                                  credentials.getId(),
+                                                  this.unixNodeType);
+            FilePath gitTempFile = gitScript.write(credentials, unbindTempDir.getDirPath());
+            credMap.put("GIT_ASKPASS", gitTempFile.getRemote());
+            return new MultiEnvironment(credMap, unbindTempDir.getUnbinder());
+        } else {
+            taskListener.getLogger().println("JGit and JGitApache type Git tools are not supported by this binding");
+            return new MultiEnvironment(credMap);
+        }
+    }
+
+    @Override
+    public Set<String> variables(@Nonnull Run<?, ?> build) {
+        Set<String> keys = new LinkedHashSet<>();
+        keys.add(GIT_USERNAME_KEY);
+        keys.add(GIT_PASSWORD_KEY);
+        return keys;
+    }
+
+    @Override
+    public void setKeyBindings(@NonNull StandardCredentials credentials) {
+        credMap.put(GIT_USERNAME_KEY, ((StandardUsernamePasswordCredentials) credentials).getUsername());
+        credMap.put(GIT_PASSWORD_KEY, ((StandardUsernamePasswordCredentials) credentials).getPassword().getPlainText());
+    }
+
+    @Override
+    public void setRunEnvironmentVariables(@NonNull FilePath filePath, @NonNull TaskListener listener) throws IOException, InterruptedException {
+        if (unixNodeType && ((CliGitAPIImpl) getGitClientInstance(gitToolExe,null,
```

Comment:
Doesn't this risk a class cast exception if the getGitClientInstance returns a JGitAPIImpl instance?

Reply:
As mentioned in this [comment](https://github.com/jenkinsci/git-plugin/pull/1104#issuecomment-877800361) the method `getCliGitTool` is the key to prevent this risk.
Also the reason to create getCliGitTool API was to ensure that the binding will only support CLI Git and no other git impls.